slidechain/export.go

```diff
@@ -271,117 +345,140 @@ func BuildExportTx(ctx context.Context, asset xdr.Asset, amount, inputAmt int64,
+	b.PushdataInt64(0).Op(op.Split).PushdataInt64(1).Op(op.Roll).Op(op.Put) // con stack: sigcheck, zeroval; arg stack: json, retireval
+	b.Tuple(func(tup *txvmutil.TupleBuilder) { tup.PushdataBytes(pubkey) }) // con stack: sigcheck, zeroval; arg stack: json, retireval, {pubkey}
+	b.Op(op.Put)                                                            // con stack: sigchecker, zeroval; arg stack: json, retireval, {pubkey}
+	b.PushdataBytes(exportContract1Prog)                                    // con stack: sigchecker, zeroval, exportContract; arg stack: json, retireval, {pubkey}
+	b.Op(op.Contract).Op(op.Call)                                           // con stack: sigchecker, zeroval
+	b.Op(op.Finalize)                                                       // con stack: sigchecker
+	prog1 := b.Build()
+	vm, err := txvm.Validate(prog1, 3, math.MaxInt64, txvm.StopAfterFinalize)
 	if err != nil {
-		return nil, errors.Wrap(err, "signing tx")
-	}
-	tx, err := tpl.Tx()
+		return nil, errors.Wrap(err, "computing transaction ID")
+	}
+	sigProg := standard.VerifyTxID(vm.TxID)
+	msg := append(sigProg, anchor...)
+	sig := ed25519.Sign(prv, msg)
+	b.PushdataBytes(sig).Op(op.Put)
+	b.PushdataBytes(sigProg).Op(op.Put)
+	b.Op(op.Call)
+
+	prog2 := b.Build()
+	log.Printf("%x", prog2)
+	tx, err := bc.NewTx(prog2, 3, math.MaxInt64)
 	if err != nil {
-		return nil, errors.Wrap(err, "building tx")
-	}
-	if inputAmt > amount {
-		txresult := txresult.New(tx)
-		output := txresult.Outputs[0].Value
-		log.Printf("output: assetid %x amount %x anchor %x", output.AssetID.Bytes(), output.Amount, output.Anchor)
+		return nil, errors.Wrap(err, "making pre-export tx")
 	}
 	return tx, nil
 }
 
 // IsExportTx returns whether or not a txvm transaction matches the slidechain export tx format.
 //
 // Expected log is:
-// For an export that fully consumes the input:
+// If input and export amounts are equal,
 // {"I", ...}
 // {"L", ...}
-// {"X", vm seed, inputAmount, asset id, anchor}
-// {"L", vm seed, refdata}
-// {"R", ...} timerange
-// {"L", ...}
+// {"O", caller, outputid}
 // {"F", ...}
 //
-// For an export that partially consumes the input:
+// Else,
 // {"I", ...}
+// {"L", ..., refdata}
 // {"L", ...}
-// {"X", vm seed, inputAmount, asset id, anchor}
-// {"L", vm seed, refdata}
-// {"L", ...}
-// {"L", ...}
+// {"O", ...}
 // {"O", caller, outputid}
-// {"R", ...}
-// {"L", ...}
 // {"F", ...}
-func IsExportTx(tx *bc.Tx, asset xdr.Asset, inputAmt int64, temp, exporter string, seqnum int64) bool {
-	// The export transaction when we export the full input amount has seven operations, and when we export
-	// part of the input and output the rest back to the exporter, it has ten operations
-	if len(tx.Log) != 7 && len(tx.Log) != 10 {
+func IsExportTx(tx *bc.Tx, asset xdr.Asset, inputAmt int64, temp, exporter string, seqnum int64, anchor, pubkey []byte) bool {
```

Comment:
It looks like this function isn't needed - it's only called in a test.

(If you want to keep calling it from that test, you could move its definition into the test file.)